src/Transports.AspNetCore/AuthorizationVisitorBase.Validation.cs

```diff
@@ -0,0 +1,214 @@
+        info.Context.ReportError(err);
+    }
+
+    /// <summary>
+    /// Adds a error to the validation context indicating that the user is not a member of any of
+    /// the roles required by this graph, field or query argument.
+    /// </summary>
+    /// <param name="info">Information about the node being validated.</param>
+    /// <param name="roles">The list of roles of which the user must be a member.</param>
+    protected virtual void HandleNodeNotInRoles(ValidationInfo info, List<string> roles)
+    {
+        var resource = GenerateResourceDescription(info);
+        var err = info.Node == null ? new AccessDeniedError(resource) : new AccessDeniedError(resource, info.Context.Document.Source, info.Node);
+        err.RolesRequired = roles;
+        info.Context.ReportError(err);
+    }
+
+    /// <summary>
+    /// Adds a error to the validation context indicating that the user is not a member of any of
+    /// the roles required by this graph, field or query argument.
+    /// </summary>
+    /// <param name="info">Information about the node being validated.</param>
+    /// <param name="policy">The policy which these nodes are being authenticated against.</param>
+    /// <param name="authorizationResult">The result of the authentication request.</param>
+    protected virtual void HandleNodeNotInPolicy(ValidationInfo info, string policy, AuthorizationResult authorizationResult)
+    {
+        var resource = GenerateResourceDescription(info);
+        var err = info.Node == null ? new AccessDeniedError(resource) : new AccessDeniedError(resource, info.Context.Document.Source, info.Node);
+        err.PolicyRequired = policy;
+        err.PolicyAuthorizationResult = authorizationResult;
+        info.Context.ReportError(err);
+    }
+
+    /// <summary>
+    /// Generates a friendly name for a specified graph, field or query argument.
+    /// </summary>
+    protected virtual string GenerateResourceDescription(ValidationInfo info)
+    {
+        if (info.Obj is ISchema)
+        {
+            return "schema";
+        }
+        else if (info.Obj is IGraphType graphType)
+        {
+            if (info.Node is GraphQLField)
+            {
+                return $"type '{graphType.Name}' for field '{info.ParentFieldType?.Name}' on type '{info.ParentGraphType?.Name}'";
+            }
+            else if (info.Node is GraphQLOperationDefinition op)
+            {
+                return $"type '{graphType.Name}' for {op.Operation.ToString().ToLower(System.Globalization.CultureInfo.InvariantCulture)} operation{(!string.IsNullOrEmpty(op.Name?.StringValue) ? $" '{op.Name}'" : null)}";
+            }
+            else
+            {
+                return $"type '{graphType.Name}'";
+            }
+        }
+        else if (info.Obj is IFieldType fieldType)
+        {
+            return $"field '{fieldType.Name}' on type '{info.ParentGraphType?.Name}'";
```

Comment:
Consider to use more precise schema coordinates syntax https://github.com/graphql/graphql-spec/pull/794

Reply:
#835